### PR TITLE
Examples for t.dispatchEvent

### PR DESCRIPTION
**Pull request by DIRECTcut**

### CODE REVIEW

File: examples/dispatch-custom-event/README.md [hunk 1]
```diff
@@ -0,0 +1,32 @@
+# Create Data-Driven Tests
```
Comment (miherlosev) [inline examples/dispatch-custom-event/README.md:1]: `# Fire custom event`

File: examples/dispatch-custom-event/README.md [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+
+This example shows how to fire [events](https://developer.mozilla.org/en-US/docs/Web/API/Event) on the webpage with TestCafe.
+
+When a [touchstart event](https://developer.mozilla.org/en-US/docs/Web/Events/touchstart) is triggered on a target item, the page dislays the event's information:
+
+* event's constructor
+* whether the event is [cancelable](https://developer.mozilla.org/en-US/docs/Web/API/Event/cancelable)
+* whether the event [bubbles](https://developer.mozilla.org/en-US/docs/Web/API/Event/bubbles)  
+
+During the test the `t.dispatchEvent` fires a `touchstart` event on the target.
```
Comment (miherlosev) [inline examples/dispatch-custom-event/README.md:15]: Move this sentence under the head.

File: examples/dispatch-custom-event/index.html [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+    <p id="target">Waiting for an event...</p>
+
+    <p> Event type: <span id="eventType"></span></p>
+
+    <p>Is cancelable: <span id="isCancelable"></span></p>
+
+    <p>Bubbles: <span id="isBubbling"></span></p>
+    
+</body>
+<script defer>
```
Comment (miherlosev) [inline examples/dispatch-custom-event/index.html:21]: Remove the `defer` attribute here and below.

File: examples/dispatch-custom-event/index.html [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+<head>
+    <meta charset="UTF-8">
+    <meta http-equiv="X-UA-Compatible" content="IE=edge">
+    <meta name="viewport" content="width=device-width, initial-scale=1.0">
+    <title>Document</title>
+</head>
+
+<body>
+    <p id="target">Waiting for an event...</p>
+
```
Comment (miherlosev) [inline examples/dispatch-custom-event/index.html:13]: Remove the new line symbols.

File: examples/dispatch-custom-event/index.html [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+    <p>Bubbles: <span id="isBubbling"></span></p>
+    
+</body>
+<script defer>
+    const target       = document.getElementById('target');
+    const eventType    = document.getElementById('eventType');
+    const isCancelable = document.getElementById('isCancelable');
+    const isBubbling   = document.getElementById('isBubbling');
+
+    window.addEventListener('touchstart', (e) => {
```
Comment (miherlosev) [inline examples/dispatch-custom-event/index.html:27]: `window.addEventListener('touchstart', e => {`

File: examples/dispatch-custom-event/index.html [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+        target.innerText       = 'Event triggered!';
+        eventType.innerText    = Object.getPrototypeOf(e)[Symbol.toStringTag];
+        isCancelable.innerText = !!e.cancelable;
+        isBubbling.innerText   = !!e.bubbles;
+    });
```
Comment (miherlosev) [inline examples/dispatch-custom-event/index.html:32]: Add a new line.

File: examples/dispatch-custom-event/index.html [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+    const event = new TouchEvent('touchstart');
```
Comment (miherlosev) [inline examples/dispatch-custom-event/index.html:33]: Add a new line.

File: examples/dispatch-custom-event/index.js [hunk 8]
```diff
@@ -0,0 +1,28 @@
+import { Selector } from 'testcafe';
+
+fixture`Dispatch Events`
+    .page('./index.html');
+
+test('Dispatch a TouchEvent', async t => {
+    const target = Selector('#target');
```
Comment (miherlosev) [inline examples/dispatch-custom-event/index.js:7]: Align assignments:
```js
const target       = Selector('#target');
 const eventType   = Selector('#eventType');
const isCancelable = Selector('#isCancelable');
const isBubbling   = Selector('#isBubbling');
```

File: examples/dispatch-custom-event/index.js [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+    const eventType = Selector('#eventType');
+    const isCancelable = Selector('#isCancelable');
+    const isBubbling = Selector('#isBubbling');
+
+    const eventArgs = {
```
Comment (miherlosev) [inline examples/dispatch-custom-event/index.js:12]: Align values:
```js
cancelable: false,
bubbles:    false
```

File: examples/dispatch-mouse-event/README.md [hunk 10]
```diff
@@ -0,0 +1,24 @@
+# Create Data-Driven Tests
```
Comment (miherlosev) [inline examples/dispatch-mouse-event/README.md:1]: `# Fire mouse event`

File: examples/dispatch-mouse-event/index.html [hunk 11]
```diff
@@ -0,0 +1,34 @@
...
+<body>
+    <div>
+        <button id="target">Target button</button>
+        <label>Seconds pressed: </label>
+        <span id="timer">0</span>
+    </div>
+
+</body>
+<script defer>
+    const btn = document.getElementById('target');
```
Comment (miherlosev) [inline examples/dispatch-mouse-event/index.html:20]: Align assignments.

File: examples/dispatch-mouse-event/README.md [hunk 12]
```diff
@@ -0,0 +1,24 @@
...
+
+**Test Code**: [index.js](index.js)
+
+**Tested Page**: [index.html](index.html)
+
+This example shows how to fire [events](https://developer.mozilla.org/en-US/docs/Web/API/Event) on the webpage with TestCafe.
+
+The tested page includes a `button` and a counter. Each second that the left click is held on the button the counter increments by one.
+
+During the test the `t.dispatchEvent` fires a [mousedown event](https://developer.mozilla.org/en-US/docs/Web/API/Element/mousedown_event) on the button and then the [mouseup event](https://developer.mozilla.org/en-US/docs/Web/API/Element/mouseup_event) in five seconds. This simulates a 5 second-long left click on the button.
```
Comment (miherlosev) [inline examples/dispatch-mouse-event/README.md:11]: Move this sentence under the head.

**Review by miherlosev — CHANGES_REQUESTED** [review]

File: examples/dispatch-mouse-event/index.html (hunk 11 above)
Comment (miherlosev) [reply examples/dispatch-mouse-event/index.html:20]: It's time to add the code linting to the `testcafe-examples` repository.
Copy the linting from here - https://github.com/DevExpress/testcafe-react-selectors/blob/master/Gulpfile.js#L25

Comment (DIRECTcut) [reply examples/dispatch-mouse-event/index.html]: in a [new PR](https://github.com/DevExpress/testcafe-examples/pull/53)

**Review by miherlosev — APPROVED** [review]

**Comment by miherlosev** [comment]
Rebase and fix the possible eslint errors.
In the near future, I will publish a TestCafe version with the `t.dispatchEvent` method.

File: examples/dispatch-custom-event/README.md [hunk 13]
```diff
@@ -0,0 +1,32 @@
...
+* whether the event [bubbles](https://developer.mozilla.org/en-US/docs/Web/API/Event/bubbles)  
+
+TestCafe invokes the [CustomEvent Constructor](https://developer.mozilla.org/en-US/docs/Web/API/CustomEvent/CustomEvent) for any event that is neither a [Mouse](https://devexpress.github.io/testcafe/documentation/reference/test-api/testcontroller/dispatchevent.html#mouse-events), a [Keyboard](https://devexpress.github.io/testcafe/documentation/reference/test-api/testcontroller/dispatchevent.html#keyboard-events), an [Input](https://devexpress.github.io/testcafe/documentation/reference/test-api/testcontroller/dispatchevent.html#input-events), a [Focus](https://devexpress.github.io/testcafe/documentation/reference/test-api/testcontroller/dispatchevent.html#focus-events) nor a [Pointer](https://devexpress.github.io/testcafe/documentation/reference/test-api/testcontroller/dispatchevent.html#pointer-events) event.
+
+To tell TestCafe to use the [TouchEvent Constructor](https://developer.mozilla.org/en-US/docs/Web/API/TouchEvent/TouchEvent), the `eventConstructor` property is passed to the method's `options`.
```
Comment (titerman) [inline examples/dispatch-custom-event/README.md:19]: "To tell..." + passive voice = bad sentence. Better to show than tell.
Could you insert a code snippet that demonstrates this option below?

```suggestion
Include the `eventConstructor` option to use the [TouchEvent Constructor](https://developer.mozilla.org/en-US/docs/Web/API/TouchEvent/TouchEvent) instead.```

File: examples/dispatch-custom-event/README.md [hunk 14]
```diff
@@ -0,0 +1,32 @@
...
+
+When `touchstart` fires, the `t.expect.eql` method checks the values displayed on the page.
```
Comment (titerman) [inline examples/dispatch-custom-event/README.md:21]: why are you saying this and what do you *want* to say? can we just remove this sentence and maybe turn it into a comment? I haven't yet read the source code, and most readers that get to this point will not have read it either. We should either provide some context to make it meaningful or remove it altogether.

```suggestion
```

File: examples/dispatch-custom-event/README.md [hunk 15]
```diff
@@ -0,0 +1,32 @@
...
+
+## TestCafe API Used in This Example
```
Comment (titerman) [inline examples/dispatch-custom-event/README.md:23]:
```suggestion
## TestCafe API methods Used in This Example
```

File: examples/dispatch-mouse-event/README.md [hunk 16]
```diff
@@ -0,0 +1,24 @@
+# Fire a Mouse Event
+
+**Test Code**: [index.js](index.js)
+
+**Tested Page**: [index.html](index.html)
```
Comment (titerman) [inline examples/dispatch-mouse-event/README.md:5]: Likewise
```suggestion
**Test Source**: [index.js](index.js)

**Page Source**: [index.html](index.html)
```

File: examples/dispatch-mouse-event/README.md [hunk 17]
```diff
@@ -0,0 +1,24 @@
...
+
+This example shows how to fire [events](https://developer.mozilla.org/en-US/docs/Web/API/Event) on the webpage with TestCafe.
```
Comment (titerman) [inline examples/dispatch-mouse-event/README.md:7]:
```suggestion
This example shows how to fire mouse [events](https://developer.mozilla.org/en-US/docs/Web/API/Event) with TestCafe.
```

File: examples/dispatch-mouse-event/README.md [hunk 18]
```diff
@@ -0,0 +1,24 @@
...
+
+During the test the `t.dispatchEvent` fires a [mousedown event](https://developer.mozilla.org/en-US/docs/Web/API/Element/mousedown_event) on the button and then the [mouseup event](https://developer.mozilla.org/en-US/docs/Web/API/Element/mouseup_event) in five seconds. This simulates a 5 second-long left click on the button.
```
Comment (titerman) [inline examples/dispatch-mouse-event/README.md:9]: simplify, simplify, simplify.
also, we should not mix integers and numerals.

```suggestion
The `t.dispatchEvent` method fires a [mousedown event](https://developer.mozilla.org/en-US/docs/Web/API/Element/mousedown_event), waits for five seconds, and then fires a [mouseup event](https://developer.mozilla.org/en-US/docs/Web/API/Element/mouseup_event). This simulates a five second-long left click.
```

File: examples/dispatch-mouse-event/README.md [hunk 19]
```diff
@@ -0,0 +1,24 @@
...
+
+The tested page includes a `button` and a counter. Each second that the left click is held on the button the counter increments by one.
```
Comment (titerman) [inline examples/dispatch-mouse-event/README.md:11]:
```suggestion
The page includes a button and a counter. For each second that you hold the button, the value of the counter increases by one.
```

File: examples/dispatch-mouse-event/README.md [hunk 20]
```diff
@@ -0,0 +1,24 @@
...
+
+When `mouseup` fires (the left click is released), the `t.expect.eql` method checks the counter value.
```
Comment (titerman) [inline examples/dispatch-mouse-event/README.md:13]: if you feel the need to clarify something in brackets, chances are, you didn't explain it right the first time.
```suggestion
When the left click is released, the `t.expect.eql` method checks the counter value.
```

File: examples/dispatch-mouse-event/README.md [hunk 21]
```diff
@@ -0,0 +1,24 @@
...
+
+## TestCafe API Used in This Example
```
Comment (titerman) [inline examples/dispatch-mouse-event/README.md:15]:
```suggestion
## TestCafe API Methods Used in This Example
```

File: examples/dispatch-mouse-event/index.js [hunk 22]
```diff
@@ -0,0 +1,19 @@
+import { Selector } from 'testcafe';
+
+fixture`HTML Element`
+    .page('./index.html');
+
+test('Hold a left-click on a button for 5 seconds', async t => {
```
Comment (titerman) [inline examples/dispatch-mouse-event/index.js:6]:
```suggestion
test('Hold a button for 5 seconds', async t => {
```

**Review by titerman — CHANGES_REQUESTED** [review]

File: examples/dispatch-custom-event/README.md [hunk 23]
```diff
@@ -0,0 +1,31 @@
...
+* whether the event [bubbles](https://developer.mozilla.org/en-US/docs/Web/API/Event/bubbles)  
+
+TestCafe invokes the [CustomEvent Constructor](https://developer.mozilla.org/en-US/docs/Web/API/CustomEvent/CustomEvent) to create events, unless the event in question is a [Mouse event](https://devexpress.github.io/testcafe/documentation/reference/test-api/testcontroller/dispatchevent.html#mouse-events), a [Keyboard event](https://devexpress.github.io/testcafe/documentation/reference/test-api/testcontroller/dispatchevent.html#keyboard-events), an [Input event](https://devexpress.github.io/testcafe/documentation/reference/test-api/testcontroller/dispatchevent.html#input-events), a [Focus event](https://devexpress.github.io/testcafe/documentation/reference/test-api/testcontroller/dispatchevent.html#focus-events) or a [Pointer event](https://devexpress.github.io/testcafe/documentation/reference/test-api/testcontroller/dispatchevent.html#pointer-events).
+
+Include the `eventConstructor` option to use the [TouchEvent Constructor](https://developer.mozilla.org/en-US/docs/Web/API/TouchEvent/TouchEvent) instead.```
```
Comment (albertov05) [inline examples/dispatch-custom-event/README.md:19]: Can you clarify what you mean by "Include"? Are you enabling this option ("Enable the...")? Or do you mean "Add the `eventConstructor` option"?

File: examples/dispatch-custom-event/README.md [hunk 24]
```diff
@@ -0,0 +1,31 @@
...
+
+
+## TestCafe API methods Used in This Example
```
Comment (albertov05) [inline examples/dispatch-custom-event/README.md:22]:
```suggestion
## TestCafe API Methods Used in This Example
```

File: examples/dispatch-mouse-event/README.md [hunk 25]
```diff
@@ -0,0 +1,24 @@
+# Fire a Mouse Event
+
+**Test Source**: [index.js](index.js)
+
+**Page Source**: [index.html](index.html)
+
+This example shows how to fire mouse [events](https://developer.mozilla.org/en-US/docs/Web/API/Event) with TestCafe.
+
+The `t.dispatchEvent` method fires a [mousedown event](https://developer.mozilla.org/en-US/docs/Web/API/Element/mousedown_event), waits for five seconds, and then fires a [mouseup event](https://developer.mozilla.org/en-US/docs/Web/API/Element/mouseup_event). This simulates a five second-long left click.
```
Comment (albertov05) [inline examples/dispatch-mouse-event/README.md:9]:
```suggestion
The `t.dispatchEvent` method fires a [mousedown event](https://developer.mozilla.org/en-US/docs/Web/API/Element/mousedown_event), waits for five seconds, and then fires a [mouseup event](https://developer.mozilla.org/en-US/docs/Web/API/Element/mouseup_event). This simulates a five second left click.
```

File: examples/dispatch-custom-event/README.md (hunk 23 above)
Comment (DIRECTcut) [reply examples/dispatch-custom-event/README.md:19]: You pass the option to a method, like so:

```
t.dispatchEvent(..., '...', {eventConstructor: 'EventName})
```

Comment (DIRECTcut) [reply examples/dispatch-custom-event/README.md:19]: Does this work

```suggestion
To use the [TouchEvent Constructor](https://developer.mozilla.org/en-US/docs/Web/API/TouchEvent/TouchEvent), Pass the `eventConstructor` option to `t.dispatchEvent`.```
```

Comment (albertov05) [reply examples/dispatch-custom-event/README.md:19]: Thank you. Please apply your suggestion.

**Comment by DIRECTcut** [comment]
waits for [Devexpress/TestCafe#6116](https://github.com/DevExpress/testcafe/issues/6116)

**Comment by miherlosev** [comment]
Try to update the used TestCafe version to `1.15.0-alpha.1` and commit the PR then it as soon as it becomes green.

**Review by titerman — APPROVED** [review]

**Pull request merged**